### ISSUE TITLE
Train with flow_from_directory and fit_generator

### ISSUE THREAD

**Issue by 1165048017** [qubvel/segmentation_models#135]
I am sorry to bother you, but there is one question confused me a lot.
I tried to use `flow_from_directory`to process the input images and labels and train with `fit_generator`. But the loss changed strangely. The loss value decreased from 0.8 to 0.2x quickly in 20 training steps in the first epoch, and then the value kept around 0.2x and didn't decrease any more.
### the process and training code is :
```
data_gen_args = dict( 
                    rotation_range=180,
                    zoom_range=0.1,
                    horizontal_flip=True,
                    vertical_flip=True)
train_image_datagen = ImageDataGenerator(**data_gen_args)
train_mask_datagen = ImageDataGenerator(**data_gen_args)

seed = 1024
batch_size=8
target_size=(512,512) #original image size
crop_size=(480,480) #cropped image size
classess_num=4  #classnum: 0 for background; 1-3 for classes

# change mask (H,W) to (H,W,class)
def mask_one_hot(mask_img,n_class):
    one_hot_mask=np.zeros((mask_img.shape[0],mask_img.shape[1],n_class),dtype=mask_img.dtype)
    for j in range(n_class):
        one_hot_mask[:,:,j]=(mask_img[:,:,0]==j).astype(mask_img.dtype)
    return one_hot_mask

#crop image
def crop_generator(batch_gen,crop_size=(512,512),class_num=4):
    while True:
        batch_x,batch_y=next(batch_gen)
        crops_img=np.zeros((batch_x.shape[0],crop_size[0],crop_size[1],3))
        crops_mask=np.zeros((batch_y.shape[0],crop_size[0],crop_size[1],class_num))
        height,width=batch_x.shape[1],batch_x.shape[2]
        for i in range(batch_x.shape[0]):
            #random crop image
            x=np.random.randint(0,height-crop_size[0]+1)
            y=np.random.randint(0,width-crop_size[1]+1)
            crops_img[i]=batch_x[i,x:x+crop_size[0],y:y+crop_size[1]]
            crops_mask[i]=mask_one_hot(batch_y[i,x:x+crop_size[0],y:y+crop_size[1]],class_num)
        yield (crops_img,crops_mask)

#image and mask generator
train_image_generator = train_image_datagen.flow_from_directory(
    'Data',
    classes=['train_img'],
    class_mode=None,
    seed=seed,
    batch_size=batch_size,
    target_size=target_size,
    interpolation='bicubic')

train_mask_generator = train_mask_datagen.flow_from_directory(
    'Data',
    classes=['train_label'],
    class_mode=None,
    seed=seed,
    batch_size=batch_size,
    target_size=target_size,
    interpolation='bicubic')
train_generator = zip(train_image_generator, train_mask_generator)
#crop image
train_crops=crop_generator(train_generator,crop_size=crop_size,class_num=classess_num)

#train
model.fit_generator(
    train_crops,
    steps_per_epoch=500,
    epochs=100000,
    callbacks=[model_checkpoint]
)
```
I just wondered what went wrong. Does the missing of `preprocess_input` cause this problem or something else?

**Comment by NiklasDL** [comment]
Make separate ImageDataGenerators classes for the Images and Masks. Use the class init argument 'preprocessing_function' within the ImageDataGenerators for your Images to use 'preprocess_input' (see https://github.com/keras-team/keras/blob/ed07472bc5fc985982db355135d37059a1f887a9/keras/preprocessing/image.py#L291).
